proposals/nnnn-access-level-on-imports.md

```diff
@@ -0,0 +1,321 @@
+Without a way to enforce the intended access-level of dependencies
+it is easy to make a mistake and expose a dependency to clients by referencing it from a public declaration even if it's intended to be an implementation detail.
+
+When a dependency is less than public and other requirements are met,
+it becomes possible to skip loading the dependency when building indirect clients.
+This can speed up build times simply because the compiler needs to do less work.
+
+## Proposed solution
+
+The core of this proposal consists of extending the current access-level logic to support declaring the existing keywords on import statements and
+applying the access-level to the imported declarations.
+
+Here's an example case where a module `DatabaseAdapter` is an implementation detail of the local module.
+We don't want to expose it to clients so we mark the import as `internal`.
+The compiler then allows references to it from internal functions but reports references from public functions.
+```swift
+internal import DatabaseAdapter
+
+internal func internalFunc() -> DatabaseAdapter.Entry {...} // Ok
+public func publicFunc() -> DatabaseAdapter.Entry {...} // error: function cannot be declared public because its result uses an internal type
+```
+
+Additionally, this proposal uses the access-level declared on each import statements in all sources files composing a module to determine how module dependencies behave.
+To balance source compatibility and best practices, the proposed default import has an implicit access-level of public in Swift 5 and of internal in Swift 6 mode.
+
+## Detailed design
+
+In this section we discuss the three main language changes of this proposal:
+accept access-level keywords on import statements to declare the visibility of the imported module,
+apply that information when type-checking the source file,
+and determine when indirect clients can skip loading transitive dependencies.
+We then cover other concerns addressed by this proposal:
+the different default access-levels of imports in Swift 5 and Swift 6,
+and the relationship with other attributes on imports.
+
+### Declaring the access-level of an imported module
+
+The access-level is declared in front of the import statement using some of the
+keywords used for a declaration: `public`, `package`, `internal`, `fileprivate`, and `private`.
+
+A public dependency can be referenced from any declaration and will be visible to all clients.
+It is declared with either the `public` keyword,
+or without a keyword in Swift 5 where an import is public by default.
+
+```swift
+public import PublicDependency
+import OtherPublicDependency // Swift 5 only
+```
+
+A dependency visible only to the modules of the same package is declared with the `package` keyword.
+Only `package`, `internal`, `fileprivate` and `private` declarations can reference the imported module.
```

Comment:
Similar to above, I think we should tweak the language here a bit to make clear that we're talking about _signatures_, not the entirety of the declaration.